src/IonReader.ts

```diff
@@ -16,23 +16,139 @@ import { Decimal } from "./IonDecimal";
+export type ReaderScalarValue = null | boolean | number | Decimal | Timestamp | string | Uint8Array;
+
 /**
- * Reads a sequence of Ion values.
+ * A pull parser interface over Ion data.
+ *
+ * Generally, users will use [[next]] and [[stepIn]] and [[stepOut]]
+ * to traverse the structure of the data. The **Value** suffixed methods are
+ * used to extract the scalar data out of the current position of the [[Reader]].
+ *
+ * [[Reader]] instances start *before* a value, thus you have to
+ * invoke [[next]] to position a newly created reader on the first value.
  */
 export interface Reader {
-  booleanValue: () => boolean;
-  byteValue: () => Uint8Array;
-  decimalValue: () => Decimal;
-  depth: () => number;
-  fieldName: () => string;
-  isNull: () => boolean;
-  next: () => IonType;
-  numberValue: () => number;
-  stepIn: () => void;
-  stepOut: () => void;
-  stringValue: () => string;
-  timestampValue: () => Timestamp;
-  type: () => IonType;
-  value: () => any;
-  annotations: () => string[];//TODO implement symboltokens to replace string[] https://github.com/amzn/ion-js/issues/121
-}
+    /**
+     * Advances the reader to the next value in the stream at the current depth.
+     *
+     * @return The corresponding [[IonType]] of the value the reader moves to, or `null`
+     *  if the reader is at the end of the stream or container.
+     */
+    next(): IonType | null;
+
+    /**
+     * Steps into the container the reader is currently positioned on.  Note that this
+     * positions the reader *before* the first value within the container.
+     *
+     * @throw Error if the reader is not positioned on a container or the container [[isNull]].
+     */
+    stepIn(): void;
+
+    /**
+     * Steps out of the current container.  This is only valid when the reader is inside of a container
+     * (i.e. `depth() > 0`).  Note that the positions the reader *after* the container that was
+     * stepped out of, but *before* the next value after the container.  One should generally
+     * call [[next]] after invoking this method.
+     *
+     * @throw Error if the reader is not positioned within a container.
+     */
+    stepOut(): void;
+
+    /** Returns the depth of the reader.  This is `0` if the reader is not inside of a container. */
+    depth(): number;
+
+    // TODO implement symboltokens to replace string[] https://github.com/amzn/ion-js/issues/121
+    /**
+     * Returns the annotations for the current value.
+     *
+     * @return the empty array if the there are no annotations or the reader is not positioned on a value.
+     */
+    annotations(): string[];
+
+    /**
+     * Returns the field name of the current value.
+     *
+     * @return `null` if the reader is not positioned on a value or is in a value that has no field name.
+     */
+    fieldName(): string | null;
+
+    /** Returns true if and only if the reader is positioned on a `null` value of any type. */
```

Comment:
so when not positioned on a value, i should expect `false`?  (just checking)

Reply:
Yes